src/lib/flow/sol-flow-js.c

```diff
@@ -0,0 +1,1422 @@
+
+        duk_pop_3(duk_ctx);
+    }
+
+    duk_pop(duk_ctx); /* out value */
+
+    return true;
+}
+
+static bool
+setup_ports(struct flow_js_type *type, const char *buf, size_t len)
+{
+    struct duk_context *duk_ctx;
+
+    duk_ctx = duk_create_heap_default();
+    if (!duk_ctx) {
+        SOL_ERR("Failed to create a Duktape heap");
+        return false;
+    }
+
+    if (duk_peval_lstring(duk_ctx, buf, len) != 0) {
+        SOL_ERR("Failed to parse javascript content: %s", duk_safe_to_string(duk_ctx, -1));
+        duk_destroy_heap(duk_ctx);
+        return false;
+    }
+    duk_pop(duk_ctx); /* duk_peval_lstring() result */
+
+    duk_push_global_object(duk_ctx);
+
+    if (!duk_get_prop_string(duk_ctx, -1, "node")) {
+        SOL_ERR("'node' variable not found in javascript file.");
+        duk_destroy_heap(duk_ctx);
+        return false;
+    }
+
+    type->str_arena = sol_arena_new();
+    if (!type->str_arena) {
+        SOL_ERR("Couldn't create sol_arena.");
+        duk_destroy_heap(duk_ctx);
+        return false;
+    }
+
+    if (!setup_ports_in(duk_ctx, type->str_arena, &type->ports_in)) {
+        duk_destroy_heap(duk_ctx);
+        return false;
+    }
+
+    if (!setup_ports_out(duk_ctx, type->str_arena, &type->ports_out)) {
+        duk_destroy_heap(duk_ctx);
+        return false;
+    }
+
+    duk_destroy_heap(duk_ctx);
+    return true;
+}
+
+static void
+flow_js_type_fini(struct flow_js_type *type)
+{
+#ifdef SOL_FLOW_NODE_TYPE_DESCRIPTION_ENABLED
+    if (type->base.description)
+        free_description(type);
+#endif
+
+    if (type->str_arena)
+        sol_arena_del(type->str_arena);
+
+    sol_vector_clear(&type->ports_in);
+    sol_vector_clear(&type->ports_out);
+
+    free(type->js_content_buf);
+}
+
+static void
+flow_dispose_type(struct sol_flow_node_type *type)
+{
+    struct flow_js_type *js_type;
+
+    SOL_NULL_CHECK(type);
+
+    js_type = (struct flow_js_type *)type;
+    flow_js_type_fini(js_type);
+    free(js_type);
+}
+
+static bool
+flow_js_type_init(struct flow_js_type *type, const char *buf, size_t len)
+{
+    char *js_content_buf;
+
+    *type = (const struct flow_js_type) {
+        .base = {
+            .api_version = SOL_FLOW_NODE_TYPE_API_VERSION,
+            .data_size = sizeof(struct flow_js_data),
+            .open = flow_js_open,
+            .close = flow_js_close,
+            .get_ports_counts = flow_js_get_ports_counts,
+            .get_port_in = flow_js_get_port_in,
+            .get_port_out = flow_js_get_port_out,
+            .dispose_type = flow_dispose_type,
+        },
+    };
+
+    if (!setup_ports(type, buf, len))
+        return false;
+
+    js_content_buf = strndup(buf, len);
+    SOL_NULL_CHECK(js_content_buf, false);
+
+    type->js_content_buf = js_content_buf;
+    type->js_content_buf_len = len;
+
+#ifdef SOL_FLOW_NODE_TYPE_DESCRIPTION_ENABLED
+    if (setup_description(type) < 0)
+        SOL_WRN("Failed to setup description");
```

Comment:
Will this cause a problem if other parts of the code assume that `SOL_FLOW_NODE_TYPE_DESCRIPTION_ENABLED` is set and tries to use description in an inconsistent state?